internal/pkg/agent/application/upgrade/marker_access_common.go

```diff
@@ -10,7 +10,7 @@ import (
 )
 
 func writeMarkerFileCommon(markerFile string, markerBytes []byte, shouldFsync bool) error {
-	f, err := os.OpenFile(markerFile, os.O_WRONLY|os.O_CREATE, 0600)
+	f, err := os.OpenFile(markerFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
```

Comment:
One thing to note with this approach is that the truncate and write do not necessarily happen atomically in the same IO operation. This could create a brief period where a read would see an empty upgrade marker file.

I am unsure how likely this is, but it seems possible. Ways to avoid this are to introduce retries when reading a 0 byte marker file or writing to a tmp file and then replacing the original file.

Reply:
> or writing to a tmp file and then replacing the original file

Implemented this in https://github.com/elastic/elastic-agent/pull/3953.